### PR TITLE
Add image processing option

### PR DESCRIPTION
**Pull request by JonathanPrince**

### CODE REVIEW

File: image-processing.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  if ((inputArgs.data instanceof stream.Stream)) {
+    originalData = new stream.PassThrough()
+    data.pipe(originalData)
+    fullsize = originalData.pipe(sharp().resize(options.maxSize, options.maxSize).max().withoutEnlargement().on('info', info => options.meta = info))
+  } else {
+    originalData = data
+    fullsize = sharp(originalData).resize(options.maxSize, options.maxSize).max().withoutEnlargement().on('info', info => options.meta = info)
+  }
+
+  const originalUpload = client.upload(name, fullsize, options).then(cache.put(name, data))
```
Comment (JonForest) [inline image-processing.js]: I don't know for sure if sharp is async, but it's use of `.on` would suggest it probably is.

If so, this upload function could fire before the resize is complete.

File: image-processing.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  if (options.thumbnails && Array.isArray(options.thumbnails)) {
+    options.thumbnails.forEach(thumbOptions => {
+      const thumbname = renameThumb(name, thumbOptions)
+      const addMeta = meta => versionOptions.meta = meta
+      const versionOptions = R.clone(options)
+      versionOptions.isThumb = thumbOptions.isThumb
+      let resizedData
+
+      if (data instanceof stream.Stream) {
+        const dataCopy = new stream.PassThrough()
```
Comment (JonForest) [inline image-processing.js]: It's not clear from the Node stream docs what this does here.  
> The stream.PassThrough class is a trivial implementation of a Transform stream that simply passes the input bytes across to the output.
Are you creating a stream out of a non-stream?

**Comment by JonForest** [comment]
If you are adding the ability to make thumbnails, the readme should probably be updated to explain how that functionality is used.

File: image-processing.js (hunk 2 above)
Comment (JonathanPrince) [inline image-processing.js]: I should have commented here, the PassThrough is being used to split the stream

File: image-processing.js (hunk 1 above)
Comment (JonathanPrince) [inline image-processing.js]: would that be an issue if the stream is already flowing?

Comment (digitalsadhu) [inline image-processing.js]: @JonForest not sure if you are aware but `client.upload` can take the raw data to upload or a stream. In this case a stream. Sharp is also a writable stream. So essentially whats happening is that originalData gets piped into sharp and then inside `client.upload`, the sharp stream is then piped into aws. (I think)

**Pull request merged**